File: client/src/test/scala/stasis/test/specs/unit/client/mocks/MockServerCoreEndpointClient.scala
package stasis.test.specs.unit.client.mocks

import java.util.concurrent.atomic.AtomicInteger

import akka.stream.Materializer
import akka.stream.scaladsl.Source
import akka.util.ByteString
import akka.{Done, NotUsed}
import stasis.client.api.clients.ServerCoreEndpointClient
import stasis.core.packaging
import stasis.core.packaging.Crate
import stasis.core.routing.Node
import stasis.test.specs.unit.client.mocks.MockServerCoreEndpointClient.Statistic

import scala.concurrent.{ExecutionContext, Future}

class MockServerCoreEndpointClient(
  override val self: Node.Id,
  crates: Map[Crate.Id, ByteString],
  pushDisabled: Boolean = false,
  pullDisabled: Boolean = false
)(implicit mat: Materializer)
    extends ServerCoreEndpointClient {
  private implicit val ec: ExecutionContext = mat.executionContext

  private val stats: Map[Statistic, AtomicInteger] = Map(
    Statistic.CratePushed -> new AtomicInteger(0),
    Statistic.CratePulled -> new AtomicInteger(0)
  )

  override val server: String = "mock-core-server"

  override def push(manifest: packaging.Manifest, content: Source[ByteString, NotUsed]): Future[Done] =
    if (!pushDisabled) {
      content
        .runFold(ByteString.empty)(_ concat _)
        .map { _ =>
          stats(Statistic.CratePushed).getAndIncrement()
          Done
        }
    } else {
      Future.failed(new RuntimeException(s"[pushDisabled] is set to [true]"))
    }

  override def pull(crate: Crate.Id): Future[Option[Source[ByteString, NotUsed]]] =
    if (!pullDisabled) {
      crates.get(crate) match {
        case Some(content) =>
          stats(Statistic.CratePulled).getAndIncrement()
          Future.successful(Some(Source.single(content)))

        case None =>
          Future.successful(None)
      }
    } else {
      Future.failed(new RuntimeException(s"[pullDisabled] is set to [true]"))
    }

  def statistics: Map[Statistic, Int] = stats.view.mapValues(_.get()).toMap
}

object MockServerCoreEndpointClient {
  def apply()(implicit mat: Materializer): MockServerCoreEndpointClient =
    new MockServerCoreEndpointClient(
      self = Node.generateId(),
      crates = Map.empty
    )

  sealed trait Statistic
  object Statistic {
    case object CratePushed extends Statistic
    case object CratePulled extends Statistic
  }
}

File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/mocks/MockServerCoreEndpointClient.kt
package stasis.test.client_android.lib.mocks

import okio.Buffer
import okio.ByteString
import okio.Source
import okio.buffer
import stasis.client_android.lib.api.clients.ServerCoreEndpointClient
import stasis.client_android.lib.model.core.CrateId
import stasis.client_android.lib.model.core.Manifest
import stasis.client_android.lib.model.core.NodeId
import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger

open class MockServerCoreEndpointClient(
    override val self: NodeId,
    private val crates: Map<CrateId, ByteString>,
    private val pushDisabled: Boolean = false,
    private val pullDisabled: Boolean = false
) : ServerCoreEndpointClient {
    private val stats: Map<Statistic, AtomicInteger> = mapOf(
        Statistic.CratePushed to AtomicInteger(0),
        Statistic.CratePulled to AtomicInteger(0)
    )

    override val server: String = "mock-core-server"

    override suspend fun push(manifest: Manifest, content: Source) {
        if (!pushDisabled) {
            content.buffer().readByteArray()
            stats[Statistic.CratePushed]?.getAndIncrement()
        } else {
            throw RuntimeException("[pushDisabled] is set to [true]")
        }
    }

    override suspend fun pull(crate: CrateId): Source? {
        if (!pullDisabled) {
            return crates[crate]?.let {
                stats[Statistic.CratePulled]?.getAndIncrement()
                Buffer().write(it)
            }
        } else {
            throw RuntimeException("[pullDisabled] is set to [true]")
        }
    }

    val statistics: Map<Statistic, Int>
        get() = stats.mapValues { it.value.get() }

    sealed class Statistic {
        object CratePushed : Statistic()
        object CratePulled : Statistic()
    }

    companion object {
        operator fun invoke(): MockServerCoreEndpointClient = MockServerCoreEndpointClient(
            self = UUID.randomUUID(),
            crates = emptyMap()
        )
    }
}
